api_server.go

```diff
@@ -0,0 +1,196 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/sirupsen/logrus"
+	"html/template"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type ApiServer struct {
+	ObjectStore
+	mux *http.ServeMux
+}
+
+type APIHandlerFunc func(r *http.Request) ([]byte, int)
+
+func (x APIHandlerFunc) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+	body, code := logRequest(x, r)
+	w.WriteHeader(code)
+	w.Write(body)
+}
+
+func logRequest(handlerFunc APIHandlerFunc, r *http.Request) ([]byte, int) {
+	start := time.Now()
+	body, code := handlerFunc(r)
+
+	clientIP := strings.Split(r.Header.Get("X-Forwarded-For"), ",")[0]
+	if clientIP == "" {
+		clientIP, _, _ = net.SplitHostPort(r.RemoteAddr)
+	}
+	fields := logrus.Fields{
+		"client_ip":       clientIP,
+		"request_path":    r.URL.EscapedPath(),
+		"user_agent":      r.UserAgent(),
+		"request_method":  r.Method,
+		"request_size":    r.ContentLength,
+		"request_seconds": time.Since(start).Seconds(),
+		"status_code":     code,
+	}
+	if code < 500 {
+		logger.WithFields(fields).Info("request completed")
+	} else {
+		logger.WithFields(fields).Error("request failed")
+	}
+	return body, code
+}
+
+func (x *ApiServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if x.mux == nil {
+		x.mux = http.NewServeMux()
+		x.mux.Handle("/music_pdfs", APIHandlerFunc(x.MusicPDFsIndex))
+		x.mux.Handle("/music_pdfs/upload", APIHandlerFunc(x.MusicPDFsUpload))
+	}
+	x.mux.ServeHTTP(w, r)
+}
+
+func (x *ApiServer) MusicPDFsIndex(r *http.Request) ([]byte, int) {
+	// only accept get
+	if r.Method != http.MethodGet {
+		return nil, http.StatusMethodNotAllowed
+	}
```

Comment:
There's a decent chance we'll want HEAD as well as OPTIONS – CORS iirc requires the latter. Not 100% certain, though.

Reply:
Yeah. I'll add them as we need them.